### ISSUE TITLE
Support for linux kernel 5.0.0 and upward

### ISSUE THREAD

**Issue by amritkrs** [martinezjavier/ldd3#37]
A lot of api's seems to have changed from older kernels. By when can we expect the changes to be incorporated?

**Comment by dwalkes** [comment]
@amritkrs, @martinezjavier  I've been working on some changes in https://github.com/cu-ecen-5013/ldd3 for a course I'm teaching this semester which will support 5.0 kernels.  It's not fully compiling yet but several of the examples are working.  I plan to do a PR at the end of the semester or could do one earlier if desired.

**Comment by dwalkes** [comment]
See PR in #39

**Comment by smalinux** [comment]
@dwalkes Thank you. that's really helpful.
I tried to make this repo, and I got this error.
Could you tell me what kernel version you use, please?
Thank you.

[root@dhcppc0 ldd3]# make
for n in misc-progs misc-modules skull scull scullc scullp sculld scullv shortprint simple tty pci usb lddbus; do make -C $n || exit 1; done
make[1]: Entering directory '/root/ldd3/misc-progs'
cc -O2 -fomit-frame-pointer -Wall -I/root/linux//include    mapper.c   -o mapper
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:15:1: error: multiple storage classes in declaration specifiers
   15 | static typedef __kernel_fd_set  fd_set;
      | ^~~~~~
/root/linux//include/linux/types.h:15:33: error: ‘fd_set’ redeclared as different kind of symbol
   15 | static typedef __kernel_fd_set  fd_set;
      |                                 ^~~~~~
In file included from /usr/include/sys/types.h:179,
                 from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/select.h:70:5: note: previous declaration of ‘fd_set’ was here
   70 |   } fd_set;
      |     ^~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:16:25: error: conflicting types for ‘dev_t’
   16 | typedef __kernel_dev_t  dev_t;
      |                         ^~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:59:17: note: previous declaration of ‘dev_t’ was here
   59 | typedef __dev_t dev_t;
      |                 ^~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:20:15: error: conflicting types for ‘nlink_t’
   20 | typedef u32   nlink_t;
      |               ^~~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:74:19: note: previous declaration of ‘nlink_t’ was here
   74 | typedef __nlink_t nlink_t;
      |                   ^~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:26:26: error: conflicting types for ‘timer_t’
   26 | typedef __kernel_timer_t timer_t;
      |                          ^~~~~~~
In file included from /usr/include/sys/types.h:130,
                 from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/bits/types/timer_t.h:7:19: note: previous declaration of ‘timer_t’ was here
    7 | typedef __timer_t timer_t;
      |                   ^~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:46:26: error: conflicting types for ‘loff_t’
   46 | typedef __kernel_loff_t  loff_t;
      |                          ^~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:42:18: note: previous declaration of ‘loff_t’ was here
   42 | typedef __loff_t loff_t;
      |                  ^~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:108:15: error: conflicting types for ‘u_int64_t’
  108 | typedef u64   u_int64_t;
      |               ^~~~~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:161:20: note: previous declaration of ‘u_int64_t’ was here
  161 | typedef __uint64_t u_int64_t;
      |                    ^~~~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:109:15: error: conflicting types for ‘int64_t’
  109 | typedef s64   int64_t;
      |               ^~~~~~~
In file included from /usr/include/sys/types.h:155,
                 from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/bits/stdint-intn.h:27:19: note: previous declaration of ‘int64_t’ was here
   27 | typedef __int64_t int64_t;
      |                   ^~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:126:13: error: conflicting types for ‘blkcnt_t’
  126 | typedef u64 blkcnt_t;
      |             ^~~~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:192:20: note: previous declaration of ‘blkcnt_t’ was here
  192 | typedef __blkcnt_t blkcnt_t;  /* Type to count number of disk blocks.  */
      |                    ^~~~~~~~
make[1]: *** [<builtin>: mapper] Error 1
make[1]: Leaving directory '/root/ldd3/misc-progs'
make: *** [Makefile:10: subdirs] Error 1
[root@dhcppc0 ldd3]# 
[root@dhcppc0 ldd3]# #==============================================================================
[root@dhcppc0 ldd3]# make
for n in misc-progs misc-modules skull scull scullc scullp sculld scullv shortprint simple tty pci usb lddbus; do make -C $n || exit 1; done
make[1]: Entering directory '/root/ldd3/misc-progs'
cc -O2 -fomit-frame-pointer -Wall -I/root/linux//include    mapper.c   -o mapper
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:15:26: error: conflicting types for ‘fd_set’
   15 | typedef __kernel_fd_set  fd_set;
      |                          ^~~~~~
In file included from /usr/include/sys/types.h:179,
                 from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/select.h:70:5: note: previous declaration of ‘fd_set’ was here
   70 |   } fd_set;
      |     ^~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:16:25: error: conflicting types for ‘dev_t’
   16 | typedef __kernel_dev_t  dev_t;
      |                         ^~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:59:17: note: previous declaration of ‘dev_t’ was here
   59 | typedef __dev_t dev_t;
      |                 ^~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:20:15: error: conflicting types for ‘nlink_t’
   20 | typedef u32   nlink_t;
      |               ^~~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:74:19: note: previous declaration of ‘nlink_t’ was here
   74 | typedef __nlink_t nlink_t;
      |                   ^~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:26:26: error: conflicting types for ‘timer_t’
   26 | typedef __kernel_timer_t timer_t;
      |                          ^~~~~~~
In file included from /usr/include/sys/types.h:130,
                 from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/bits/types/timer_t.h:7:19: note: previous declaration of ‘timer_t’ was here
    7 | typedef __timer_t timer_t;
      |                   ^~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:46:26: error: conflicting types for ‘loff_t’
   46 | typedef __kernel_loff_t  loff_t;
      |                          ^~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:42:18: note: previous declaration of ‘loff_t’ was here
   42 | typedef __loff_t loff_t;
      |                  ^~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:108:15: error: conflicting types for ‘u_int64_t’
  108 | typedef u64   u_int64_t;
      |               ^~~~~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:161:20: note: previous declaration of ‘u_int64_t’ was here
  161 | typedef __uint64_t u_int64_t;
      |                    ^~~~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:109:15: error: conflicting types for ‘int64_t’
  109 | typedef s64   int64_t;
      |               ^~~~~~~
In file included from /usr/include/sys/types.h:155,
                 from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/bits/stdint-intn.h:27:19: note: previous declaration of ‘int64_t’ was here
   27 | typedef __int64_t int64_t;
      |                   ^~~~~~~
In file included from /root/linux//include/linux/limits.h:6,
                 from /usr/include/bits/local_lim.h:38,
                 from /usr/include/bits/posix1_lim.h:161,
                 from /usr/include/limits.h:183,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:194,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-redhat-linux/9/include/limits.h:34,
                 from mapper.c:27:
/root/linux//include/linux/types.h:126:13: error: conflicting types for ‘blkcnt_t’
  126 | typedef u64 blkcnt_t;
      |             ^~~~~~~~
In file included from /usr/include/stdlib.h:394,
                 from mapper.c:22:
/usr/include/sys/types.h:192:20: note: previous declaration of ‘blkcnt_t’ was here
  192 | typedef __blkcnt_t blkcnt_t;  /* Type to count number of disk blocks.  */
      |                    ^~~~~~~~
make[1]: *** [<builtin>: mapper] Error 1
make[1]: Leaving directory '/root/ldd3/misc-progs'
make: *** [Makefile:10: subdirs] Error 1

**Comment by dwalkes** [comment]
Thanks @smalinux 

> Could you tell me what kernel version you use, please?

```
Ubuntu 18.04 kernel as of January 2020: 5.3.0-42-generic
Yocto warrior branch kernel for qemu builds: 5.0.19
Buildroot 2019.05 kernel for qemu builds: 4.9.16
```

I've added this to the [README](https://github.com/cu-ecen-5013/ldd3/blob/master/README)

I haven't seen any responses to https://github.com/martinezjavier/ldd3/pull/39 so I'll create my own parallel project repository after the semester ends.  You can open issues/PRs against this one and I'll accept them over the summer as I prepare for the next semester.

**Comment by dwalkes** [comment]
I believe this issue should be closed with #39 please re-open if not.
